### ISSUE TITLE
feat: adaptive-stencil mask dispatch for non-advection operators

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/finitevolX#205]
## Summary

Follow-up to jejjohnson/finitevolX#204.  PR #204 wires `ArakawaCGridMask` into every class operator using the **post-compute zero by output stagger** convention.  That convention is correct for everything except a small set of operators where the stencil reads neighbour cells whose values are not yet zeroed in dry cells, leading to subtly wrong tendencies in the wet cell adjacent to land.  This issue tracks upgrading those operators to **adaptive-stencil dispatch** — the trick `Advection2D` already uses today via `ArakawaCGridMask.get_adaptive_masks()` and `upwind_flux`.

**Parent epic:** #29 — finitevolX upstream — masks everywhere + missing operators
**Upstream PR:** #204 — masks-everywhere refactor (must merge first)
**Related somax issue:** jejjohnson/somax#82

## Background — why post-compute zero is sometimes insufficient

The convention used by PR #204 is the simplest possible: compute the operator as if the domain were all-ocean, then multiply the output by `mask.<output_stagger>`.  Concretely, for the Laplacian on a coastal T-cell:

```python
out = interior((diff_x_fwd(h) - diff_x_bwd(h)) / dx**2 +
               (diff_y_fwd(h) - diff_y_bwd(h)) / dy**2, h)
out = out * mask.h   # zero the dry T-cells
```

The issue: at a wet T-cell `(j, i)` whose east neighbour `(j, i+1)` is dry, the second-order stencil still reads `h[j, i+1]` (which is whatever the user put there — typically zero, sometimes NaN, sometimes garbage from a previous integrator step).  Multiplying the wet-cell output by 1 doesn't fix the contamination — the wet cell's tendency was already polluted before the multiply.

In practice this is fine when the input field is itself mask-clean (e.g. produced by a previous mask-aware operator that already zeroed dry cells), and PR #204's golden tests pass on the canonical 18×18 cross-shaped-island fixture.  But for accuracy-sensitive work near complex coastlines — or for callers that pass raw external data with non-zero values in dry cells — we want a more defensive approach.

## What "adaptive-stencil dispatch" looks like

`ArakawaCGridMask` already exposes the machinery:

```python
adaptive = mask.get_adaptive_masks(direction='x', source='h')
# adaptive[2]  -> 1st-order upwind only
# adaptive[4]  -> WENO3   / 3-point centred safe
# adaptive[6]  -> WENO5   / 5-point centred safe
# adaptive[8]  -> WENO7
# adaptive[10] -> WENO9
```

Each entry is a mutually-exclusive boolean mask telling you the largest stencil width that fits at each cell without reading a dry neighbour.  `Advection2D` already uses this via `upwind_flux` to blend reconstruction orders cell-by-cell.  The same pattern can be applied to the post-compute-zero operators: at `distbound1` cells, fall back to a 1-point stencil; at `distbound2` cells, fall back to the standard 3-point central difference; at `distbound3plus` cells, use the full operator.

For the centred-difference operators in this repo, the practical "fall back" usually means: at coastal cells, use a one-sided difference instead of the centred stencil.

## Operators that would benefit

All centred-stencil class operators that read across faces.  In approximate priority order:

| Operator | Method | Reads dry-side neighbours when … |
|---|---|---|
| `Difference2D/3D` | `laplacian` | wet T-cell adjacent to dry T-cell (5-point cross) |
| `Difference2D/3D` | `divergence` | wet T-cell adjacent to dry U/V faces |
| `Difference2D` | `curl` | corner with mixed wet/dry adjacent faces |
| `Difference2D` | `grad_perp` | psi reads 4-point stencil across diagonal neighbours |
| `Vorticity2D/3D` | `relative_vorticity`, `potential_vorticity` | same as `curl` |
| `Interpolation2D/3D` | `T_to_X`, `X_to_T`, `U_to_V`, `V_to_U` | 4-point bilinear with mixed wet/dry corners |
| `SphericalDifference2D/3D` | `diff2_lon`, `laplacian_merid` | same Laplacian-style stencil widening |
| `SphericalLaplacian2D/3D` | `__call__` | same |
| `BiharmonicDiffusion2D/3D` | `__call__` | the *intermediate* Laplacian (already documented as a known caveat) |

`Diffusion2D/3D` already does the right thing via the inline intermediate-flux masking added in PR #204; it does not need this upgrade.

## Proposed approach

Mirror what `Advection2D` does:

1. Define a small set of fallback stencils per operator (1-point, 3-point central, etc.).  For most of the operators above, the fallback at `distbound1` cells is "use the existing finitevolX one-sided helper" and at `distbound2+` cells is "use the standard centred stencil unchanged."
2. At call time, `mask.get_adaptive_masks(direction=…, source=…)` returns mutually-exclusive boolean tier masks.  The operator computes each tier separately and `jnp.where`'s the results together.
3. The post-compute multiply by `mask.<output_stagger>` stays — it's still needed to zero dry cells where no fallback is sensible.
4. The unmasked code path is unchanged (fast path: just the existing centred stencil).

The dispatch adds a constant-factor overhead in the masked path; for most production runs that's worth it for correctness near complex coastlines.

## Acceptance criteria

- [ ] Pick a representative operator (probably `Difference2D.laplacian`) and prototype the adaptive-stencil dispatch end-to-end
- [ ] Demonstrate on a coastal-mask test that the new dispatch eliminates the wet-cell contamination from dry-side neighbour reads (vs. PR #204's post-compute zero)
- [ ] Generalise the dispatch helper so it can be applied to the other centred-stencil operators in the table without per-operator boilerplate
- [ ] Migrate `Difference2D/3D.{laplacian, divergence, curl, grad_perp}` to the new dispatch
- [ ] Migrate `Vorticity2D/3D` and the spherical Laplacian / Laplacian-merid operators
- [ ] Migrate `Interpolation2D/3D` 4-point stencils (`T_to_X`, `X_to_T`, `U_to_V`, `V_to_U`)
- [ ] Add per-operator regression tests that compare adaptive-dispatch output against PR #204's post-compute-zero output on an all-ocean mask (must be bit-identical) and on a coastal mask (must differ exactly on the wet cells immediately adjacent to land)
- [ ] Document the upgrade in `docs/masks.md` "Operator API" section as the now-default convention, demoting post-compute zero to "fast path / fallback"

## Out of scope

- Adaptive dispatch for the *spherical compound* operators (`SphericalDivergence2D/3D`, `SphericalVorticity2D/3D`).  Those should be added in a separate pass to keep this issue scoped.
- Extending `BiharmonicDiffusion` beyond the existing post-compute multiply.  The intermediate-Laplacian Dirichlet-0 caveat is documented in the class docstring; truly fixing it requires a full mask-aware Laplacian intermediate, which is a different ticket.
- Performance benchmarking of the masked path.  Track separately if it becomes load-bearing.
